sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4

```diff
@@ -42,6 +42,29 @@ options { tokenVocab = SqlBaseLexer; }
   public boolean double_quoted_identifiers = false;
 }
 
+compoundOrSingleStatement
+    : singleStatement
+    | singleCompound
+    ;
+
+singleCompound
+    : compound SEMICOLON* EOF
+    ;
+
+compound
+    : BEGIN compoundBody END
+    ;
+
+// Last semicolon in body is optional.
+compoundBody
```

Comment:
is there any value in having this be a separate rule? Can we just list the above as the following instead?

```
    : BEGIN compoundStatement (SEMICOLON+ compoundStatement)* SEMICOLON* END
```

Reply:
By separating these 2 rules we can reuse `compoundBody` inside `if/else`, `while` and some other things that we are planning to add

Reply:
As Milan said, we will reuse `compoundBody` in future in the blocks that don't require `BEGIN` and `END` (if else, while, etc.).
One more reason is that in such blocks (that don't begin with `BEGIN`) variable declaration is not allowed, so we need to have these cases separated.

Reply:
Sounds good, thanks for explaining